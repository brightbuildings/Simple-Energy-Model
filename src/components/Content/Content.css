.Content {
  background-color: white;
  padding: 2rem;
  display: flex;
  justify-content: space-between;
}

.Navigation {
  width: 25%;
}

.Navigation button {
  margin-bottom: 1rem;
}

.Navigation button.active {
  font-weight: bold;
}

.subNavigation {
  display: flex-start;
  justify-content: center;
  margin-bottom: 2rem;
}

.subNavigation button {
  padding: 1rem;
}

.subNavigation button.active {
  font-weight: bold;
}

.Fields {
  width: 75%;
  padding-right: 4rem;
}

.Fields h2 {
  margin-top: 0;
}

.fieldset {
  display: none;
}

.fieldset.active {
  display: initial;
}

.input-group {
  margin-bottom: 1.5rem;
}

label {
  width: 100%;
  margin-bottom: 0;
  font-weight: 600;
  margin-left: 0.75rem;
}

.spacer {
  height: 40px;
}