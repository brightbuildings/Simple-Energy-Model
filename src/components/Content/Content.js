import React, { useState } from "react";
import DataService from "../../services/DataService";
import { FormGroup, Label, Input, InputGroup, InputGroupAddon, InputGroupText } from "reactstrap";

import "./Content.css";

const formatSections = (data, activeSection, activeSubsection, setActiveSubsection, variables, setVariables) => {
  if (data == null) {
    return;
  }
  return Object.entries(data).map(section => {
    const key = section[0];
    const value = section[1];

    // If Parameters, export both A and B
    if (key === "parameters") {
      const parametersA = [key+"-a", value];
      const parametersB = [key+'-b', value];

      return (
        <React.Fragment key="parameters">
          {formatSection(parametersA, activeSection, activeSubsection, setActiveSubsection, variables, setVariables)}
          {formatSection(parametersB, activeSection, activeSubsection, setActiveSubsection, variables, setVariables)}
        </React.Fragment>
      );
    } else {
      return formatSection(section, activeSection, activeSubsection, setActiveSubsection, variables, setVariables);
    }
  });
};

const formatSection = (section, activeSection, activeSubsection, setActiveSubsection, variables, setVariables) => {
  const key = section[0];
  const value = section[1];

  let labelKey = "label";
  let contentKey = "content";
  if (key === "parameters-b") {
    labelKey = "labelB";
    contentKey = "contentB";
  }
  return (
    <FormGroup key={key} className={`fieldset ${activeSection === key ? "active" : ""}`}>
      <h2>{value[labelKey]}</h2>
      <p>{value[contentKey]}</p>

      <div className="subNavigation">
        {Object.entries(value.fields).map((field, index) => {
          const key = field[0];
          const navTitle = field[1].navigation;
          console.log(field);

          let activeKey = activeSubsection;
          if (index === 0 && activeSubsection === "") {
            activeKey = key;
          }

          return (
            <button 
              key={key}
              onClick={() => setActiveSubsection(key)}
              className={key === activeKey ? "active" : "" } 
            >
              {navTitle}
            </button>
          );
        })}
      </div>
      <div className="subNavigationContent">
        {Object.entries(value.fields).map((field, index) => {
          const key = field[0];
          const value = field[1];

          if (index === 0 && activeSubsection === "") {
            // display this entry
          } else if (activeSubsection !== key) {
            return null;
          }

          return (
            <section key={key} className="Questions">
              {Object.entries(value.fields).map(question => {
                return formatQuestion(question, variables, setVariables);
              })}
            </section>
            );
          })
        }
      </div>
    </FormGroup>
  );
}

const formatNavigation = (data, activeSection, setActiveSection, setActiveSubsection) => {
  if (data == null) {
    return;
  }
  return Object.entries(data).map(section => {
    let key = section[0];

    // If Parameters, export both A and B
    if (key === "parameters") {
      return (
        <React.Fragment key="parameters">
          <button 
            className={`${section[0]}-a` === activeSection ? "active" : "" } 
            key={section[0]+"a"}
            onClick={() => {
              setActiveSubsection("");
              setActiveSection(section[0]+"-a")
            }}
          >
            Parameters A
          </button>
          <br />
          <button 
            className={`${section[0]}-b` === activeSection ? "active" : "" } 
            key={section[0]+"b"} 
            onClick={() => {
              setActiveSubsection("");
              setActiveSection(section[0]+"-b")
            }}
          >
            Parameters B
          </button>
          <br />
        </React.Fragment>
      )
    } else {
      return (
        <React.Fragment key={section[0]}>
          <button
            className={section[0] === activeSection ? "active" : "" } 
            onClick={() => {
              setActiveSubsection("");
              setActiveSection(section[0]);
            }}
          >
            {section[0]}
          </button>
          <br />
        </React.Fragment>
      );
    }
  });
};

const formatQuestion = (question, variables, setVariables) => {
  if (question == null) {
    return;
  }
  const key = question[0];
  const value = question[1];

  let required = false;
  if (value?.required === true) {
    required = true;
  }

  switch (value?.type) {
    case "float": // fall through
    case "integer":
      return (
        <InputGroup key={key}>
          <Label for={key}>{value.label}</Label>
          <Input 
            type="number" 
            required={required}
            name={key}
            id={key}
            defaultValue={variables[key]}
            onChange={input => setVariables({...variables, [key]: input.currentTarget.value})}
          />
          {value.unit && (
            <InputGroupAddon addonType="append">
              <InputGroupText>{value.unit}</InputGroupText>
            </InputGroupAddon>
          )}
        </InputGroup>
      );
    case "text":
      return (
        <InputGroup key={key}>
          <Label for={key}>{value.label}</Label>
          <Input 
            type="text" 
            required={required}
            name={key}
            id={key}
            defaultValue={variables[key]}
            onChange={input => setVariables({...variables, [key]: input.currentTarget.value})}
          />
          {value.unit && (
            <InputGroupAddon addonType="append">
              <InputGroupText>{value.unit}</InputGroupText>
            </InputGroupAddon>
          )}
        </InputGroup>
      );
    case "gap":
      return <div key={key} className="spacer">&nbsp;</div>;
    default:
      return;
  }
};

function Content() {
  const [activeSection, setActiveSection] = useState("introduction");
  const [activeSubsection, setActiveSubsection] = useState(null);
  const [variables, setVariables] = useState({});
  const data = DataService.load();

  return (
    <div className="centered-narrow">
      <div className="Content">
        <div className="Fields">
          {formatSections(data, activeSection, activeSubsection, setActiveSubsection, variables, setVariables)}
        </div>
        <div className="Navigation">
          {formatNavigation(data, activeSection, setActiveSection, setActiveSubsection)}
        </div>
      </div>
    </div>
  );
}

export {
  Content,
  formatSections,
  formatNavigation,
  formatQuestion
};
