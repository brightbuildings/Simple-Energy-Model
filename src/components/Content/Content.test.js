import { Content, formatSections, formatNavigation, formatQuestion } from './Content';
import renderer from "react-test-renderer";

const json = {
  "introduction": {
    "label": "Test Label",
    "content": "Test Content",
    "fields": {}
  }
};

const questionFloat = [
  "questionFloat",
  {
    "required": true,
    "type": "float",
    "label": "Float",
  }
];

const questionInteger = [
  "question2Key",
  {
    "required": false,
    "type": "integer",
    "label": "Integer",
    "unit": "IBU"
  }
];

const questionText = [
  "question3Key",
  {
    "required": true,
    "type": "text",
    "label": "Text Label"
  }
];

const questionGap = [
  "question4Key",
  {
    "type": "gap"
  }
];

test("runs formatSection with empty json", () => {
  expect(formatSections()).toBeUndefined();
});

test("runs formatSection with json", () => {
  const result = formatSections(json);
  expect(result).not.toBeUndefined();
});

test("runs formatNavigation with empty json", () => {
  expect(formatSections()).toBeUndefined();
});

test("runs formatNavigation with json", () => {
  const result = formatNavigation(json);
  expect(result).not.toBeUndefined();
});

test("runs formatQuestion with empty question", () => {
  expect(formatQuestion()).toBeUndefined();
});

test("runs formatQuestion with float question", () => {
  const result = formatQuestion(questionFloat);
  expect(result).not.toBeUndefined();
});

test("runs formatQuestion with integer question", () => {
  const result = formatQuestion(questionInteger);
  expect(result).not.toBeUndefined();
});

test("runs formatQuestion with text question", () => {
  const result = formatQuestion(questionText);
  expect(result).not.toBeUndefined();
});

test("runs formatQuestion with gap question", () => {
  const result = formatQuestion(questionGap);
  expect(result).not.toBeUndefined();
});

test("renders formatQuestion with float question", () => {
  const result = renderer.create(formatQuestion(questionFloat));
  expect(result).toMatchSnapshot();
});

test("renders formatQuestion with integer question", () => {
  const result = renderer.create(formatQuestion(questionInteger));
  expect(result).toMatchSnapshot();
});

test("renders formatQuestion with text question", () => {
  const result = renderer.create(formatQuestion(questionText));
  expect(result).toMatchSnapshot();
});

test("renders formatQuestion with gap question", () => {
  const result = renderer.create(formatQuestion(questionGap));
  expect(result).toMatchSnapshot();
});

test("renders Content", () => {
  const tree = renderer.create(<Content />).toJSON();
  expect(tree).toMatchSnapshot();
});
