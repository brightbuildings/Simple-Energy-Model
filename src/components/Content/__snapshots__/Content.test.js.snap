// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders Content 1`] = `
<div
  className="centered-narrow"
>
  <div
    className="Content"
  >
    <div
      className="Fields"
    >
      <div
        className="fieldset active form-group"
      >
        <h2>
          Introduction
        </h2>
        <p>
          Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.
        </p>
        <div
          className="subNavigation"
        >
          <button
            className=""
            onClick={[Function]}
          >
            introduction
          </button>
          <button
            className=""
            onClick={[Function]}
          >
            dimensions
          </button>
          <button
            className=""
            onClick={[Function]}
          >
            solar
          </button>
          <button
            className=""
            onClick={[Function]}
          >
            location
          </button>
        </div>
        <div
          className="subNavigationContent"
        />
      </div>
      <div
        className="fieldset  form-group"
      >
        <h2>
          Parameters - Model A
        </h2>
        <p>
          How is the existing home built?
        </p>
        <div
          className="subNavigation"
        >
          <button
            className=""
            onClick={[Function]}
          >
            build
          </button>
          <button
            className=""
            onClick={[Function]}
          >
            mechanical
          </button>
          <button
            className=""
            onClick={[Function]}
          >
            energyPrices
          </button>
        </div>
        <div
          className="subNavigationContent"
        />
      </div>
      <div
        className="fieldset  form-group"
      >
        <h2>
          Parameters - Model B
        </h2>
        <p>
          What improvements would you like to consider?
        </p>
        <div
          className="subNavigation"
        >
          <button
            className=""
            onClick={[Function]}
          >
            build
          </button>
          <button
            className=""
            onClick={[Function]}
          >
            mechanical
          </button>
          <button
            className=""
            onClick={[Function]}
          >
            energyPrices
          </button>
        </div>
        <div
          className="subNavigationContent"
        />
      </div>
      <div
        className="fieldset  form-group"
      >
        <h2>
          Financing the Net-zero Upgrade
        </h2>
        <p>
          Lorem ipsum
        </p>
        <div
          className="subNavigation"
        >
          <button
            className=""
            onClick={[Function]}
          >
            improvementsSection1
          </button>
          <button
            className=""
            onClick={[Function]}
          >
            improvementsSection2
          </button>
          <button
            className=""
            onClick={[Function]}
          >
            improvementsSection3
          </button>
        </div>
        <div
          className="subNavigationContent"
        />
      </div>
      <div
        className="fieldset  form-group"
      >
        <h2>
          Business Case
        </h2>
        <p>
          Lorem ipsum
        </p>
        <div
          className="subNavigation"
        >
          <button
            className=""
            onClick={[Function]}
          >
            paceLoanTerm
          </button>
          <button
            className=""
            onClick={[Function]}
          >
            interest
          </button>
        </div>
        <div
          className="subNavigationContent"
        />
      </div>
    </div>
    <div
      className="Navigation"
    >
      <button
        className="active"
        onClick={[Function]}
      >
        introduction
      </button>
      <br />
      <button
        className=""
        onClick={[Function]}
      >
        Parameters A
      </button>
      <br />
      <button
        className=""
        onClick={[Function]}
      >
        Parameters B
      </button>
      <br />
      <button
        className=""
        onClick={[Function]}
      >
        financing
      </button>
      <br />
      <button
        className=""
        onClick={[Function]}
      >
        business
      </button>
      <br />
    </div>
  </div>
</div>
`;

exports[`renders formatQuestion with float question 1`] = `
<div
  className="input-group"
>
  <label
    className=""
    htmlFor="questionFloat"
  >
    Float
  </label>
  <input
    className="form-control"
    required={true}
    type="number"
  />
</div>
`;

exports[`renders formatQuestion with gap question 1`] = `
<div
  className="spacer"
>
  Â 
</div>
`;

exports[`renders formatQuestion with integer question 1`] = `
<div
  className="input-group"
>
  <label
    className=""
    htmlFor="question2Key"
  >
    Integer
  </label>
  <input
    className="form-control"
    required={false}
    type="number"
  />
  <div
    className="input-group-append"
  >
    <span
      className="input-group-text"
    >
      IBU
    </span>
  </div>
</div>
`;

exports[`renders formatQuestion with text question 1`] = `
<div
  className="input-group"
>
  <label
    className=""
    htmlFor="question3Key"
  >
    Text Label
  </label>
  <input
    className="form-control"
    required={true}
    type="text"
  />
</div>
`;
